reconfiguration/include/reconfiguration/pruning_handler.hpp

```diff
@@ -17,22 +17,93 @@
+  //  PruneRequest is equal to the number of replicas in the system
+  //  - verifying the signatures of individual LatestPrunableBlock messages in the
+  //  PruneRequest.
+  // If all above conditions are met, the state machine will prune blocks from the
+  // genesis block up to the the minimum of all the block IDs in
+  // LatestPrunableBlock messages in the PruneRequest message.
  public:
-  bool handle(const concord::messages::LatestPrunableBlockRequest&, concord::messages::LatestPrunableBlock&) override {
-    return true;
-  }
-  bool handle(const concord::messages::PruneStatusRequest&, concord::messages::PruneStatus&) override { return true; }
-  bool handle(const concord::messages::PruneRequest&, kvbc::BlockId&) override { return true; }
+  // Construct by providing an interface to the storage engine, state transfer,
+  // configuration and tracing facilities. Note this constructor may throw an
+  // exception if there is an issue with the configuration (for example, if the
+  // configuration enables pruning but does not provide a purning operator
+  // public key).
+  PruningHandler(std::shared_ptr<kvbc::IReader>,
+                 std::shared_ptr<kvbc::IBlockAdder>,
+                 std::shared_ptr<kvbc::IBlocksDeleter>,
+                 std::shared_ptr<bftEngine::IStateTransfer>,
+                 bool run_async = false);
+  bool handle(const concord::messages::LatestPrunableBlockRequest &, concord::messages::LatestPrunableBlock &) override;
+  bool handle(const concord::messages::PruneRequest &, kvbc::BlockId &) override;
+  bool handle(const concord::messages::PruneStatusRequest &, concord::messages::PruneStatus &) override;
+  static std::string lastAgreedPrunableBlockIdKey() { return last_agreed_prunable_block_id_key_; }
 
- protected:
-  logging::Logger getLogger() {
-    static logging::Logger logger_(logging::getLogger("concord.pruning"));
-    return logger_;
-  }
-};
+ private:
+  kvbc::BlockId latestBasedOnNumBlocksConfig() const;
+  kvbc::BlockId latestBasedOnTimeRangeConfig() const;
 
-}  // namespace concord::reconfiguration
+  kvbc::BlockId agreedPrunableBlockId(const concord::messages::PruneRequest &) const;
+  // Returns the last agreed prunable block ID from storage, if existing.
+  std::optional<kvbc::BlockId> lastAgreedPrunableBlockId() const;
+  void persistLastAgreedPrunableBlockId(kvbc::BlockId block_id) const;
+  // Prune blocks in the [genesis, block_id] range (both inclusive).
+  // Throws on errors.
+  void pruneThroughBlockId(kvbc::BlockId block_id) const;
+  void pruneThroughLastAgreedBlockId() const;
+  void pruneOnStateTransferCompletion(uint64_t checkpoint_number) const noexcept;
+  logging::Logger logger_;
+  RSAPruningSigner signer_;
+  RSAPruningVerifier verifier_;
+  std::shared_ptr<kvbc::IReader> ro_storage_;
+  std::shared_ptr<kvbc::IBlockAdder> blocks_adder_;
+  std::shared_ptr<kvbc::IBlocksDeleter> blocks_deleter_;
+
+  bool pruning_enabled_{false};
```

Comment:
If we don't keep the configuration as local class variables, but always ask ReplicaConfig::instance().get() it would make future dynamic configuration impl easier.